Fix reduceItemByOne and its tests

ChrisBrownie55 opened the pull request:
This fixes #48 where `reduceItemByOne` doesn't cause a re-render. However, we should probably update our tests to take a more _"does this render to the screen correctly"_ instead of _"has this property been modified/replaced to the correct value."_

I haven't verified that his has fixed the issue, however, it should have, I'll check once I get the chance. In the meantime, what do y'all think the best way to rewrite the tests to ensure the correct rendering would be?

dayhaysoos commented:
@ChrisBrownie55 I'm gonna test the rerendering right now. There's gotta be a way to test if there's a rerender event some how, I'm going to look into some options

dayhaysoos commented:
@ChrisBrownie55 it works!!!

I actually don't quite understand why it works. Mind walking me through what's happening here? I'm very interested lol

ChrisBrownie55 commented:
So, the way the code was written before was that the `cart` was one big object that had a property called `cartItems`. This meant that for the screen to update and re-render you needed to return a new object when something changed (it didn't matter what data you changed inside the object). React only cares about referential equality, not the structure of the variable.

When I changed the code to memoize the value of `cartItems` to combine it with the rest of the `cart`, the code became reliant on the fact that a new `cartItems` array was being created (just like how we were relying on a new `cart` object to be created). So, when `Array#splice` is called, we don't generate a new array of any sort, instead the array is modified in-place. Due to this in-place modification, React can't detect any changes to `cartItems` and it just assumes that it's the same `cartItems` due to referential equality.

To fix this, we just needed a reference to a new/different array. This is why I rewrote the function to do the following:

```js
const newCartItems = [];
// Add stuff to it
return newCartItems;
```

dayhaysoos commented:
Thank you @ChrisBrownie55, looks great to me. I think I wanna merge this now so I can make sure it doesn't conflict with other PRs

ChrisBrownie55 commented:
@dayhaysoos sounds good to me, we can perhaps look into better tests at a later point 👍
